gitclone/steperror.go

```diff
@@ -1,11 +1,178 @@
+
+func mapRecommendation(tag string, err error) step.Recommendation {
+	msg := err.Error()
+	switch tag {
+	case checkoutFailedTag:
+		matcher := newCheckoutFailedPatternErrorMatcher()
+		return matcher.Run(msg)
+	case updateSubmodelFailedTag: // update_submodule_failed could have the same errors as fetch
+		fallthrough
+	case fetchFailedTag:
+		fetchFailedMatcher := newFetchFailedPatternErrorMatcher()
+		return fetchFailedMatcher.Run(msg)
+	}
+	return nil
+}
 
 func newStepError(tag string, err error, shortMsg string) *step.Error {
+	recommendation := mapRecommendation(tag, err)
+	if recommendation != nil {
+		return step.NewErrorWithRecommendations("git-clone", tag, err, shortMsg, recommendation)
+	}
+
 	return step.NewError("git-clone", tag, err, shortMsg)
 }
 
+func newDetailedErrorRecommendation(detail Detail) step.Recommendation {
+	return step.Recommendation{
+		"DetailedError": map[string]string{
+			"Title":       detail.Title,
+			"Description": detail.Description,
+		},
+	}
+}
+
+func checkoutFailedGenericDetail(params ...string) Detail {
+	err := params[0]
+	return Detail{
+		Title:       "We couldn’t checkout your branch.",
+		Description: fmt.Sprintf("Our auto-configurator returned the following error:\n%s", err),
+	}
+}
+
+func fetchFailedGenericDetail(params ...string) Detail {
+	err := params[0]
+	return Detail{
+		Title:       "We couldn’t fetch your repository.",
+		Description: fmt.Sprintf("Our auto-configurator returned the following error:\n%s", err),
+	}
+}
+
+func fetchFailedSSHAccessError(params ...string) Detail {
+	return Detail{
+		Title:       "We couldn’t access your repository.",
+		Description: "Please abort the process, double-check your SSH key and try again.",
+	}
+}
+
+func fetchFailedCouldNotFindGitRepository(params ...string) Detail {
+	repoURL := params[0]
+	return Detail{
+		Title:       fmt.Sprintf("We couldn’t find a git repository at '%s'.", repoURL),
+		Description: "Please abort the process, double-check your repository URL and try again.",
+	}
+}
+
+func fetchFailedHTTPAccessError(params ...string) Detail {
+	return Detail{
+		Title:       "We couldn’t access your repository.",
+		Description: "Please abort the process and try again, by providing the repository with SSH URL.",
+	}
+}
+
+func fetchFailedCouldConnectError(params ...string) Detail {
+	host := params[0]
+	return Detail{
+		Title:       fmt.Sprintf("We couldn’t connect to '%s'.", host),
+		Description: "Please abort the process, double-check your repository URL and try again.",
+	}
+}
+
+func fetchFailedSamlSSOError(params ...string) Detail {
+	return Detail{
+		Title:       "To access this repository, you need to use SAML SSO.",
+		Description: `Please abort the process, update your SSH settings and try again. You can find out more about <a target="_blank" href="https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/authorizing-an-ssh-key-for-use-with-saml-single-sign-on">using SAML SSO in the Github docs</a>.`,
```

Comment:
is our frontend prepared to present HTML content properly?

Reply:
Not sure, but we will have to change it anyway to support the new DetailedError structure, so we can make sure to handle it.